### PR TITLE
(BIDS-2224) Do a double check for machine offline

### PR DESCRIPTION
**Pull request by Eisei24**
### <samp>🤖 Generated by Copilot at c9692e4</samp>

Improved the logic and flexibility of machine offline notifications in `services/notifications.go`. Introduced a variable to allow for a higher tolerance of offline ratio for the first check, and handled various cases of machine status and user settings.

### CODE REVIEW

File: services/notifications.go [hunk 1]
```diff
@@ -2342,11 +2344,37 @@ func collectMonitoringMachine(
...
+		`SELECT 
+			COUNT(DISTINCT user_id)
+			FROM users_subscriptions
+			WHERE event_name = $1`,
+		eventName)
+	if err != nil {
+		return err
+	}
+
+	// If less than 10 users are subscribed send the notifications
```
[the remark this answers is not in the document]
Comment (Eisei24) [reply services/notifications.go:2362]: Changed. (with "few")

File: services/notifications.go [hunk 2]
```diff
@@ -2342,11 +2344,37 @@ func collectMonitoringMachine(
...
+		}
+		if float64(len(result))/float64(len(allSubscribed)) >= subRatioThreshold {
+			utils.LogError(nil, fmt.Errorf("error too many users would be notified concerning: %v", eventName), 0)
+			return nil
+		}
+	}
+	if eventName == types.MonitoringMachineOfflineEventName {
+		// Notifications will be sent, reset the flag
+		isFirstNotificationCheck = true
 	}
```
[the remark this answers is not in the document]
Comment (Eisei24) [reply services/notifications.go:2378]: I moved the block to the end of the function but I wouldn't change the comment.
Notifications still haven't been sent.

File: services/notifications.go [hunk 3]
```diff
@@ -2342,11 +2344,37 @@ func collectMonitoringMachine(
...
 
-	// if at least 90% of users would be notified, we expect an issue on our end and no one will be notified
-	const notifiedSubscriptionsRatioThreshold = 0.9
-	if float64(len(result))/float64(len(allSubscribed)) >= notifiedSubscriptionsRatioThreshold {
-		utils.LogError(nil, fmt.Errorf("error too many users would be notified concerning: %v", eventName), 0)
-		return nil
+	const subThreshold = 10
+	const subFirstRatioThreshold = 0.3
+	const subSecondRatioThreshold = 0.9
```
[the remark this answers is not in the document]
Comment (Eisei24) [reply services/notifications.go:2349]: I remember that he reviewed the original
`const notifiedSubscriptionsRatioThreshold = 0.9`
and was fine with it then.

[the remark this answers is not in the document]
Comment (Eisei24) [reply services/notifications.go]: As requested the three threshold were added to the config.
The behaviour was kept like the `OnlineDetectionLimit` notification config, the default value is given in the code but can be overridden by adding the yaml keyword to the used `XXX.config.yml` file.

**Review by D13ce — APPROVED** [review]
lgtm

**Pull request merged**